academy/ibc/ibc-tao-dev.md

```diff
@@ -0,0 +1,314 @@
+
+<insert image here>
+
+In the code, you can find the reference implementation for the connection handshake in the [IBC module repository](https://github.com/cosmos/ibc-go/blob/main/modules/core/03-connection/keeper/handshake.go). In it take a look at `ConnOpenInit`:
+
+```go
+func (k Keeper) ConnOpenInit(
+  ctx sdk.Context,
+  clientID string,
+  counterparty types.Counterparty, // counterpartyPrefix, counterpartyClientIdentifier
+  version *types.Version,
+  delayPeriod uint64,
+) (string, error) {
+  
+  ...
+
+  // connection defines chain A's ConnectionEnd
+  connectionID := k.GenerateConnectionIdentifier(ctx)
+  connection := types.NewConnectionEnd(types.INIT, clientID, counterparty, types.ExportedVersionsToProto(versions), delayPeriod)
+  k.SetConnection(ctx, connectionID, connection)
+
+  if err := k.addConnectionToClient(ctx, clientID, connectionID); err != nil {
+    return "", err
+  }
+
+  k.Logger(ctx).Info("connection state updated", "connection-id", connectionID, "previous-state", "NONE", "new-state", "INIT")
+
+  defer func() {
+    telemetry.IncrCounter(1, "ibc", "connection", "open-init")
+  }()
+
+  EmitConnectionOpenInitEvent(ctx, connectionID, clientID, counterparty)
+
+  return connectionID, nil
+}
+
+```
+
+This function creates an unique `connectionID`. You can see that it adds the connection to a list, which represents the connections associated with a specific client.
+
+Also you can see how it creates a new `ConnectionEnd`:
+
+```go
+connection := types.NewConnectionEnd(types.INIT, clientID, counterparty, types.ExportedVersionsToProto(versions), delayPeriod)
+````
+
+`ConnOpenInit` is triggered by the **relayer**, which constructs the message and hands it over to the SDK that uses the [`msg_server.go`](https://github.com/cosmos/ibc-go/blob/main/modules/core/keeper/msg_server.go) you saw before to call `ConnOpenInit`:
+
+```go
+
+// ConnectionOpenInit defines a rpc handler method for MsgConnectionOpenInit.
+func (k Keeper) ConnectionOpenInit(goCtx context.Context, msg *connectiontypes.MsgConnectionOpenInit) (*connectiontypes.MsgConnectionOpenInitResponse, error) {
+  ctx := sdk.UnwrapSDKContext(goCtx)
+
+  if _, err := k.ConnectionKeeper.ConnOpenInit(ctx, msg.ClientId, msg.Counterparty, msg.Version, msg.DelayPeriod); err != nil {
+    return nil, sdkerrors.Wrap(err, "connection handshake open init failed")
+  }
+
+  return &connectiontypes.MsgConnectionOpenInitResponse{}, nil
+}
+```
+
+<HighlightBox type="info">
+
+In IBC, blockchains do not directly pass messages to each other over the network. To communicate, blockchains commit the state to a precisely defined path reserved for a specific message type and a specific counterparty. In the case of establishing a connection,  Relayers monitor for updates on these paths and relay messages by submitting the data stored under the path along with proof of that data to the counterparty chain.
```

Comment:
should this be in here? looks a different tone than the rest

Reply:
will reference in another section